lib/elixir/lib/enum.ex

```diff
@@ -1542,18 +1542,25 @@ defmodule Enum do
   ## Examples
 
       # Although not necessary, let's seed the random algorithm
-      iex> :rand.seed(:exsplus, {1, 2, 3})
-      iex> Enum.random([1, 2, 3])
-      2
+      iex> :rand.seed(:exsplus, {101, 102, 103})
       iex> Enum.random([1, 2, 3])
       1
+      iex> Enum.random([1, 2, 3])
+      3
 
   """
   @spec random(t) :: element | no_return
+  def random(enumerable)
+
+  def random(first..last),
+    do: random_integer(first, last)
+
   def random(enumerable) do
-    case take_random(enumerable, 1) do
-      [] -> raise Enum.EmptyError
-      [e] -> e
+    case count(enumerable) do
+      0 ->
+        raise Enum.EmptyError
+      count ->
+        at(enumerable, random_integer(0, count - 1))
```

Comment:
@josevalim that's pretty ingenious how to not to traverse enumerable twice.
code update, we are good to merge